### PR TITLE
[Draft] Add module user_role

### PR DESCRIPTION
**Pull request by muehlings**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

There is no module for creating, editing or deleting user roles.

## What is the new behavior?

[Draft] Module for creating, editing or deleting user roles.

## Other information

First try of generating Ansible modules from OpenAPI definition for Checkmk. This code is 99% generated from the checkmk.yaml file in SwaggerUI. Manually changed parts are around the [...] brackets. Module is based on Checkmk 2.2.0p20-CME YAML.

### CODE REVIEW

**Comment by muehlings** [comment]
Playbook:
```
  - name: "Create user_role '{{item.ident}}"
    checkmk.general.user_role:
      "automation_secret": "{{automation_secret}}"
      "automation_user": "{{automation_user}}"
      "server_url": "{{server_url}}"
      "site": "{{site}}"
      "state": "{{item.state}}"

      "new_permissions": "{{item.new_permissions}}"
      "role_id": "{{item.role_id}}"
      "new_alias": "{{item.new_alias}}"
      "new_role_id": "{{item.new_role_id}}"
      "new_basedon": "{{item.new_basedon}}"
    loop: "{{ user_roles|flatten(levels=1) }}"
    loop_control:
      label: "{{ item.role_id }}"

```

host_vars:
```
---
user_roles:
- new_alias: Normal monitoring user (copy2)
  new_basedon: user
  new_role_id: userx
  role_id: userx
  new_permissions:
  - general.use
  - general.see_all
  - view.host_pending
  - view.problemsofhost
  - view.hostgroup
    [...]
  - view.hostgroup_up
  - notification_plugin.asciimail
  - notification_plugin.opsgenie_issues
  - notification_plugin.sms
  - notification_plugin.pagerduty
  state: present
  title: Normal monitoring user (copy)
```

File: plugins/modules/user_role.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+# i.e. the version is of the form "2.5.0" and not "2.4".
+version_added: "0.11.0"
+
+description:
+- Manage user_role in Checkmk.
+
+extends_documentation_fragment: [checkmk.general.common]
+
+options:
+    [...]
```
Comment (robin-checkmk) [inline plugins/modules/user_role.py:27]: Documentation is a must-have. Please align the documentation section with existing modules and the `options` section with the `module_args`.

File: plugins/modules/user_role.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+author:
+    - Stefan Mühling (@muehlings)
+"""
+
+EXAMPLES = r"""
+    [...]
```
Comment (robin-checkmk) [inline plugins/modules/user_role.py:34]: Please provide at least some basic examples.

**Review by robin-checkmk — CHANGES_REQUESTED** [review]

File: .github/workflows/ans-int-test-user_role.yaml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+# - When changing the module name, it needs to be changed in 'env:MODULE_NAME' and in 'on:pull_requests:path'!
+#
+# Resources:
+# - Template for this file: https://github.com/ansible-collections/collection_template/blob/main/.github/workflows/ansible-test.yml
+# - About Ansible integration tests: https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html
+
+env:
+  NAMESPACE: checkmk
+  COLLECTION_NAME: general
+  MODULE_NAME: user_role
```
Comment (robin-checkmk) [inline .github/workflows/ans-int-test-user_role.yaml:11]: We need to discuss the name of the module before merging it.

**Comment by lgetwan** [comment]
Hi @muehlings,

I created integration tests for your module. As I expected, they are failing, as I'm not quite sure what parameters are used for what. To me it looks like the parameters change their meaning, depending on the current state of the user_role. When creating it, "user_role" means the existing role to be cloned, but in the 2nd run, when the role already exists, it means the name of the created rule.

I know that's caused by the parameters of the REST API, but in the Ansible module, we should make it more user friendly. That's what our collection is about. :-)

I suggest (independent from the current state of the roles in Checkmk):

- name: String that results in the name/role_id,
- title: String that results in the title/alias,
- based_on: String that set's the existing role that our role should be based from,
- permissions: List of strings that represent the permissions that are explicitly set for this role.
 
Once, the parameters are defined, please make your module idempotent and integrate documentation and examples.
You also have to adjust the integration test to fit to what you defined. Please adjust these files accordingly:

- tests/integration/targets/user_role/vars/main.yml 
- tests/integration/targets/user_role/tasks/test.yml

Then, we will review and merge it.

**Pull request closed without merging**